frontend/storm_wiki/util/search.py

```diff
@@ -0,0 +1,171 @@
+                    category,
+                    set(id_str.split("_(")[0] for id_str in processed_ids),
+                )
+
+        except FileNotFoundError:
+            logger.warning(
+                "Wikipedia sources file not found. Domain restrictions will not be applied."
+            )
+        except IOError as e:
+            logger.error(f"Error reading Wikipedia sources file: {e}")
+        except Exception as e:
+            logger.error(f"Unexpected error in _generate_domain_restriction: {e}")
+
+    def _is_valid_wikipedia_source(self, url):
+        parsed_url = urlparse(url)
+        domain = parsed_url.netloc.split(".")[-2]  # Get the domain name
+        combined_set = self.generally_unreliable | self.deprecated | self.blacklisted
+        return domain not in combined_set
+
+    def forward(
+        self, query_or_queries: Union[str, List[str]], exclude_urls: List[str] = []
+    ) -> List[Dict[str, Any]]:
+        if not dspy.settings.rm:
+            raise ValueError("No RM is loaded. Please load a Retrieval Model first.")
+        return super().forward(query_or_queries, exclude_urls=exclude_urls)
+
+
+class CombinedSearchAPI(WebSearchAPIWrapper):
+    def __init__(self, max_results=20):
+        super().__init__(max_results)
+        self.ddg_search = DuckDuckGoSearchAPIWrapper()
+        self.searxng_base_url = os.getenv("SEARXNG_BASE_URL", "http://localhost:8080")
+
+    def _rank_results(self, results: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
+        # Implement your ranking logic here
+        # For example, you could rank based on the presence of certain keywords,
+        # the length of the snippet, or the domain authority of the source
+        return sorted(results, key=lambda x: self._calculate_relevance(x), reverse=True)
+
+    def _calculate_relevance(self, result: Dict[str, Any]) -> float:
+        # Implement your relevance calculation here
+        # This is a simple example; you might want to use more sophisticated methods
+        relevance = 0.0
+        if "wikipedia.org" in result["url"]:
+            relevance += 1.0
+        relevance += len(result["snippets"][0]) / 1000  # Favor longer snippets
+        return relevance
+
+    def forward(
+        self, query_or_queries: Union[str, List[str]], exclude_urls: List[str] = []
+    ) -> List[Dict[str, Any]]:
+        queries = (
+            [query_or_queries]
+            if isinstance(query_or_queries, str)
+            else query_or_queries
+        )
+        all_results = []
+
+        for query in queries:
+            try:
+                # Try DuckDuckGo first
+                results = self._search_duckduckgo(query)
+            except Exception as e:
+                logger.warning(
+                    f"DuckDuckGo search failed: {str(e)}. Falling back to SearxNG."
+                )
+                try:
+                    # Fallback to SearxNG
+                    results = self._search_searxng(query)
```

Comment:
we can change this flow once we have more search engine options. and let the user to choose which engine to use.